File: python_client/src/globe.py
"""
globe.py holds the key methods for the WebSocket API.
"""
import json
from datetime import datetime, timezone
import hmac
import hashlib
import base64
from uuid import uuid4
import websockets
import aiohttp

# pylint: disable=R0904


class Globe:
    """
    Globe class contains all of Globes WebSocket WS channel subscriptions
    """

    def __init__(self, error_handler=None, authentication=None):
        self.uri = "wss://globedx.com/api/v1/ws"
        self.socket = websockets
        self.http_api = "http://www.globedx.com/api/v1"
        self.error_handler = error_handler
        self.received_handlers = {}
        self.session = aiohttp.ClientSession()
        self.authentication = authentication

    async def _send(self, message):
        """
        Send the subscription channel request to the server.
        """
        await self.socket.send(json.dumps(message))

    async def run_loop(self):
        """
        Continually listen and process any responses on the websocket from the server.
        """
        while True:
            received = await self.socket.recv()
            received = json.loads(received)
            if "subscription" in received:
                if "instrument" in received["subscription"]:
                    key = "{}{}".format(
                        received["subscription"]["channel"],
                        received["subscription"]["instrument"])
                    if key in self.received_handlers:
                        await self.received_handlers[key](
                            received
                        )
                elif received["subscription"]["channel"] in self.received_handlers:
                    await self.received_handlers[received["subscription"]["channel"]](
                        received)
                else:
                    print(received)
            else:
                if self.error_handler:
                    self.error_handler(received)
                else:
                    print("Globe Error: ", received)

    async def get_depth(self, instrument, handler=None):
        """
        Subscribe to the the market depth (l2 orderbook) with a given handler, for an instrument.
        """
        if handler is not None:
            key = "depth{}".format(instrument)
            self.received_handlers[key] = handler
        message = {"command": "subscribe",
                   "channel": "depth", "instrument": instrument}
        await self._send(message)

    async def get_product_list(self, handler=None):
        """
        Get the product list with a given handler
        """
        if handler is not None:
            self.received_handlers["product-list"] = handler
        message = {"command": "subscribe", "channel": "product-list"}
        await self._send(message)

    async def get_index_price(self, instrument, handler=None):
        """
        Subscribe to the index price with a given handler, for an instrument.
        """
        if handler is not None:
            key = "index-price{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "index-price",
            "instrument": instrument,
        }
        await self._send(message)

    async def get_product_detail(self, instrument, handler=None):
        """
        Get the product detail with a given handler, for an instrument.
        """
        if handler is not None:
            key = "product-detail{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "product-detail",
            "instrument": instrument,
        }
        await self._send(message)

    async def get_recent_trades(self, instrument, handler=None):
        """
        Subscribe to recent trades with a given handler, for an instrument.
        """
        if handler is not None:
            key = "trades{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "trades",
            "instrument": instrument,
        }
        await self._send(message)

    async def get_market_overview(self, instrument, handler=None):
        """
        Get the market overview with a given handler, for an instrument.
        """
        if handler is not None:
            key = "market-overview{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "market-overview",
            "instrument": instrument,
        }
        await self._send(message)

    async def get_price_history(self, instrument, resolution="1m", handler=None):
        """
        Subscribe to the price history with a given handler, for an instrument and resolution.
        """
        if handler is not None:
            key = "price-history{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "price-history",
            "instrument": instrument,
            "resolution": resolution,
        }
        await self._send(message)

    async def get_open_interest(self, instrument, handler=None):
        """
        Gets the open interest with a given handler, for an instrument.
        """
        if handler is not None:
            key = "open-interest{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "open-interest",
            "instrument": instrument,
        }
        await self._send(message)

    async def get_insurance_fund(self, handler=None):
        """
        Get the insurance fund, with a given handler, for an instrument.
        """
        if handler is not None:
            self.received_handlers["insurance-fund"] = handler
        message = {
            "command": "subscribe",
            "channel": "insurance-fund",
        }
        await self._send(message)

    async def get_my_account_overview(self, handler=None):
        """
        Get your account overview, with a given handler.
        """
        if handler:
            self.received_handlers["my-account-overview"] = handler
        message = {"command": "subscribe", "channel": "my-account-overview"}
        await self._send(message)

    async def get_my_market_events(self, instrument, handler=None):
        """
        Subscribe to your market events, with a given handler.
        """
        if handler is not None:
            key = "my-market-events{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "my-market-events",
            "instrument": instrument,
        }
        await self._send(message)

    async def cancel_order(self, _id, instrument, cancel_id=None, new_quantity=None) -> str:
        """
        Submit an order cancellation, with a given order id and instrument.
        """
        message = {"command": "cancel-order",
                   "instrument": instrument, "order_id": _id}
        if cancel_id:
            message["cancel_id"] = cancel_id
        else:
            message["cancel_id"] = str(uuid4())
        if new_quantity:
            message["new_quantity"] = new_quantity
        await self._send(message)
        return message["cancel_id"]

    async def cancel_stop_order(self, _id, instrument, cancel_id=None) -> str:
        """
        Submit an stop order cancellation, with a given order id and instrument.
        """
        message = {"command": "cancel-stop-order",
                   "instrument": instrument, "order_id": _id}
        if cancel_id:
            message["cancel_id"] = cancel_id
        else:
            message["cancel_id"] = str(uuid4())
        await self._send(message)
        return message["cancel_id"]

    async def my_open_orders(self, instrument, handler=None):
        """
        Get your open orders, with a given handler, for an instrument.
        """
        if handler is not None:
            key = "my-orders{}".format(instrument)
            self.received_handlers[key] = handler
        message = {
            "command": "subscribe",
            "channel": "my-orders",
            "instrument": instrument,
        }
        await self._send(message)

    async def my_positions(self, handler=None):
        """
        Get your positions, with a given handler.
        """
        if handler:
            self.received_handlers["my-positions"] = handler
        message = {"command": "subscribe", "channel": "my-positions"}
        await self._send(message)

    async def place_order(self, order) -> str:
        """
        Place an order.
        """
        if 'order_id' not in order:
            order['order_id'] = str(uuid4())
        order["command"] = "place-order"
        await self._send(order)
        return order['order_id']

    async def get_historic_market_rates(self, instrument, resolution):
        """
        Get historic OHLC bars for a instrument and are returned
        in grouped buckets based on requested resolution.
        """
        endpoint = (
            self.http_api + "/history/" + instrument + "/candles/" + resolution
        )
        async with self.session.get(endpoint) as output:
            output = await output.json()
        return output

    async def get_open_orders(self, instrument, upto_timestamp=None, page_size=100):
        """
        Get your current open orders for a product.
        """
        endpoint = (
            self.http_api + "/orders/open-orders"
        )
        if not upto_timestamp:
            upto_timestamp = int(datetime.now().timestamp() * 1000)

        params = {
            'instrument': instrument,
            'upto_timestamp': upto_timestamp,
            'page_size': page_size,
        }
        extra_headers = self.auth_headers(url="GET/api/v1/orders/open-orders")
        async with self.session.get(endpoint, params=params, headers=extra_headers) as output:
            output = await output.text()
        return output

    async def get_positions(self):
        """
        Get your current positions for instruments.
        """
        endpoint = (
            self.http_api + "/positions"
        )

        extra_headers = self.auth_headers(url="GET/api/v1/positions")
        async with self.session.get(endpoint, headers=extra_headers) as output:
            output = await output.json()
        return output

    async def get_account_overview(self):
        """
        Get your current account overview.
        """
        endpoint = (
            self.http_api + "/account-overview"
        )

        extra_headers = self.auth_headers(url="GET/api/v1/account-overview")
        async with self.session.get(endpoint, headers=extra_headers) as output:
            output = await output.json()
        return output

    async def get_my_trades(self, instrument, page=None):
        """
        Get your current my previous trades for a product.
        """
        endpoint = (
            self.http_api + "/history/my-trades"
        )
        if not page:
            page = 0

        params = {
            'instrument': instrument,
            'page': page,
        }
        extra_headers = self.auth_headers(url="GET/api/v1/history/my-trades")
        async with self.session.get(endpoint, params=params, headers=extra_headers) as output:
            output = await output.text()
        return output

    async def get_historic_index_price_rates(self, instrument, resolution):
        """
        Get historic OHLC bars for an indexp price of an instrument
        and are returned in grouped buckets based on requested resolution.
        """
        endpoint = (
            self.http_api
            + "/history/index-price/"
            + instrument
            + "/candles/"
            + resolution
        )
        async with self.session.get(endpoint) as output:
            output = await output.json()
        return output

    def auth_headers(self, url):
        """
        Generate new authentication headers
        """
        if not self.authentication:
            raise Exception("Authentication is required")
        headers = {
            "X-Access-Key": self.authentication["api-key"],
            "X-Access-Signature": "",
            "X-Access-Nonce": str(int(datetime.now(timezone.utc).timestamp() * 1000)),
            "X-Access-Passphrase": self.authentication["passphrase"],
        }
        secret = self.authentication["secret"]
        sign_txt = headers["X-Access-Nonce"] + url
        headers["X-Access-Signature"] = str(
            _hash(sign_txt, secret), "utf-8")
        return headers

    async def connect(self):
        """
        Connect to the websocket server.
        """
        if not self.authentication:
            self.socket = await websockets.connect(self.uri)
            print("Connected to Globe Websocket.")
        if self.authentication:
            extra_headers = self.auth_headers(url="GET/api/v1/ws")
            self.socket = await websockets.connect(self.uri, extra_headers=extra_headers)
            print("Connected to Globe Websocket.")


def _hash(sign_txt, secret):
    """
    Hashing function for the signature and authentication.
    """
    sign_txt = bytes(sign_txt, "utf-8")
    secret = bytes(secret, "utf-8")
    secret = base64.b64decode(secret)
    signature = base64.b64encode(
        hmac.new(secret, sign_txt, digestmod=hashlib.sha256).digest()
    )
    return signature

File: typescript-node/src/globe.ts
import Websocket from "ws"
import crypto from "crypto"
import fetch from "node-fetch"
import { v4 as uuidv4 } from "uuid"

export enum Side {
  Buy = "buy",
  Sell = "sell",
}

export type ResolutionType =
  | "1m"
  | "3m"
  | "5m"
  | "15m"
  | "30m"
  | "1h"
  | "2h"
  | "4h"
  | "6h"
  | "12h"
  | "1d"
  | "3d"
  | "1w"

export enum OrderType {
  Limit = "limit",
  PostOnly = "post-only",
  Market = "market",
}

export enum OrderExecutionType {
  Stop = "stop",
  Order = "order",
}

type RequestId = string

export type OpenOrder = {
  type: OrderExecutionType
  filled_quantity: number
  order_id: RequestId
  order_type: OrderType
  price: number
  quantity: number
  side: Side
  timestamp: number
}

export type Order = {
  instrument: string
  order_type: OrderType
  quantity: number
  price?: number
  side: Side
  order_id?: RequestId
}

export type StopOrder = Order & {
  trigger_price: number
}

export type CancelOrder = {
  instrument: string
  order_id: RequestId
  cancel_id?: RequestId
  new_quantity?: number
}

export type reduceQuantity = {
  instrument: string
  order_id: RequestId
  cancel_id?: RequestId
  new_quantity: number
}

export const Channel = {
  /**
   * Subscription message for the market depth channel.
   * Responses contain arrays of the best long and short prices for the given instrument.
   *
   * This channel does not require authentication.
   */
  depth: (instrument: string) => {
    return { instrument, channel: "depth" }
  },

  /**
   * Subscription message for the index price channel.
   * Responses contain the current index price for the given instrument.
   *
   * This channel does not require authentication.
   */
  indexPrice: (instrument: string) => {
    return { instrument, channel: "index-price" }
  },

  /**
   * Subscription message for the insurance fund channel.
   * Responses contain the current balance of the insurance fund.
   *
   * This channel does not require authentication.
   */
  insuranceFund: (instrument: string) => {
    return { instrument, channel: "insurance-fund" }
  },

  /**
   * Subscription message for product list channel.
   * Responses contain an array of the current available products.
   *
   * This channel does not require authentication.
   */
  productList: () => {
    return { channel: "product-list" }
  },

  /**
   * Subscription message for the product details channel.
   * Responses contain objects with details of the contract for the given instrument,
   * including fees and funding period.
   *
   * This channel does not require authentication.
   */
  productDetail: (instrument: string) => {
    return { instrument, channel: "product-detail" }
  },

  /**
   * Subscription message for the recent trades channel.
   * Responses contain an array of the 100 most recent trades for the given instrument.
   *
   * This channel does not require authentication.
   */
  trades: (instrument: string) => {
    return { instrument, channel: "trades" }
  },

  /**
   * Subscription message for the market overview channel.
   * Responses are objects containing 24 hour trading volume, price change and funding rate.
   *
   * This channel does not require authentication.
   */
  marketOverview: (instrument: string) => {
    return { instrument, channel: "market-overview" }
  },

  /**
   * Subscription message for the price history channel.
   * Responses are objects containing candle metrics (high, low, open, close, time and volume)
   *
   * This channel does not require authentication.
   */
  priceHistory: (instrument: string, resolution: ResolutionType) => {
    return { instrument, resolution, channel: "market-overview" }
  },

  /**
   * Subscription message for the market open interest channel.
   * Responses are the number of outstanding contracts for the given instrument.
   *
   * This channel does not require authentication.
   */
  openInterest: (instrument: string) => {
    return { instrument, channel: "open-interest" }
  },

  /**
   * Subscription message for the market events channel.
   * Responses are market data messages for the current user, including order confirmations, rejections and trades.
   *
   * This channel is personalized per trader and requires authentication.
   */
  myMarketEvents: (instrument: string) => {
    return { instrument, channel: "my-market-events" }
  },

  /**
   * Subscription message for the account overview channel.
   * Responses contain account information for the current user, including balances, margin limits and PnL.
   *
   * This channel is personalized per trader and requires authentication.
   */
  myAccountOverview: () => {
    return { channel: "my-account-overview" }
  },

  /**
   * Subscription message for the open orders channel.
   * Responses are updates to your open orders for the specified instrument.
   * The updates are objects whose keys are order ids. The values are either the order details or `null`
   * to indicate the order has been removed (by being cancelled or filled).
   *
   * This channel is personalized per trader and requires authentication.
   */
  myOpenOrders: (instrument: string) => {
    return { instrument, channel: "my-orders" }
  },

  /**
   * Subscription message for the positions channel.
   * Responses are objects whose keys are instrument symbols
   * and whose values contain the quantity and average entry price for the position.
   *
   * This channel is personalized per trader and requires authentication.
   */
  myPositions: () => {
    return { channel: "my-positions" }
  },
}

// Api types

export type ApiCredentials = {
  secret: string
  apiKey?: string
  passphrase: string
}

export type GlobeInit = {
  credentials?: ApiCredentials
  host?: string
  ssl?: boolean
  connectionTimeout?: number
}

type Fields = { [key: string]: any }
type ReceiveHandler = (event: Websocket.Data) => void
type Message = { path: string; method: string; body: string }
type ErrorHandler = (event: Websocket.Data) => void

const sign = (request: string, credentials: ApiCredentials) => {
  const nonce = new Date().getTime()
  const hmac = crypto.createHmac(
    "sha256",
    Buffer.from(credentials.secret, "base64"),
  )
  const messageText = `${nonce}${request}`

  return {
    passphrase: credentials.passphrase,
    nonce,
    signature: hmac.update(messageText).digest("base64"),
    api_key: credentials.apiKey,
  }
}

const signMessage = (msg: Message, credentials: ApiCredentials) =>
  sign(`${msg.method}${msg.path}${msg.body}`, credentials)

const authHeaders = (request: Message, credentials: ApiCredentials) => {
  const signed = signMessage(request, credentials)
  return {
    "X-Access-Key": `${signed.api_key}`,
    "X-Access-Signature": signed.signature,
    "X-Access-Nonce": `${signed.nonce}`,
    "X-Access-Passphrase": signed.passphrase,
  }
}

// Time

const withTimeout = <T>(
  timeout: number,
  promise: Promise<T>,
  message: string,
  cleanup?: () => void,
): Promise<T> =>
  Promise.race([
    promise,
    new Promise<T>((resolve, reject) => {
      setTimeout(() => {
        if (cleanup) {
          cleanup()
        }
        reject(new Error(`Timeout after ${timeout}ms. Context: ${message}`))
      }, timeout)
    }),
  ])

// Client

export class Globe {
  private error: ErrorHandler
  private ws?: Websocket
  private path: string
  private host: string
  private httpApi: string
  private connectionTimeout: number
  private receiveHandlers: Record<string, ReceiveHandler> = {}
  private credentials?: ApiCredentials

  constructor(error: ErrorHandler, init?: GlobeInit) {
    const host = init?.host || "globedx.com"
    const ssl = init?.ssl === undefined ? true : init?.ssl
    const secure = ssl ? "s" : ""
    this.error = error
    this.path = "/api/v1/ws"
    this.host = `ws${secure}://${host}`
    this.httpApi = `http${secure}://${host}/api/v1`
    this.connectionTimeout = init?.connectionTimeout || 30000
    this.credentials = init?.credentials
  }

  async connect(): Promise<Globe> {
    const path = this.path
    const headers =
      this.credentials &&
      authHeaders({ path, method: "GET", body: "" }, this.credentials)
    this.ws = new Websocket(this.host + this.path, { headers })

    this.ws.onmessage = (message) => {
      const data = JSON.parse(message.data as string)

      if (data.subscription) {
        if (this.receiveHandlers[data.subscription.channel]) {
          this.receiveHandlers[data.subscription.channel](data)
        } else if (
          data.subscription.channel === "my-orders" &&
          this.receiveHandlers[
            data.subscription.channel + "-" + data.subscription.instrument
          ]
        ) {
          this.receiveHandlers[
            data.subscription.channel + "-" + data.subscription.instrument
          ](data)
        }
      } else if (data.error) {
        this.error(data)
      }
    }

    if (this.ws.readyState && Websocket.OPEN) {
      return Promise.resolve(this)
    }

    return withTimeout(
      this.connectionTimeout,
      new Promise<Globe>((resolve, reject) => {
        if (this.ws) {
          this.ws.onerror = (error) => {
            reject(error)
          }

          this.ws.onopen = () => {
            resolve(this)
          }
        } else {
          reject(new Error("Error while creating the Websocket"))
        }
      }),
      `Connecting to '${this.path}'`,
    )
  }

  private addReceiveHandler(channel: string, handle: ReceiveHandler) {
    this.receiveHandlers[channel] = handle
  }

  private send(message: Fields) {
    if (!this.ws) {
      throw new Error("Socket not connected")
    }

    this.ws.send(JSON.stringify(message))
  }

  subscribe(
    details: { instrument?: string; channel: string },
    handle: ReceiveHandler,
  ) {
    this.send({
      command: "subscribe",
      ...details,
    })
    if (details.channel === "my-orders") {
      this.addReceiveHandler(details.channel + "-" + details.instrument, handle)
    } else {
      this.addReceiveHandler(details.channel, handle)
    }
  }

  unsubscribe(details: { instrument?: string; channel: string }) {
    this.send({
      command: "unsubscribe",
      ...details,
    })

    // Need to remove the handle here as well
  }

  close() {
    if (!this.ws) {
      console.warn("Socket not connected")
      return
    }

    if (this.ws.readyState === Websocket.OPEN) {
      this.ws.terminate()
    }
  }

  // Public REST

  async getHistoricMarketRates(
    instrument: string,
    resolution: ResolutionType,
  ): Promise<any> {
    const endpoint =
      this.httpApi + `/history/${instrument}/candles/${resolution}`
    return fetch(endpoint)
      .then((res) => res.text())
      .then((text) => JSON.parse(text))
  }

  async getHistoricIndexPriceRates(
    instrument: string,
    resolution: ResolutionType,
  ): Promise<any> {
    const endpoint =
      this.httpApi + `/history/index-price/${instrument}/candles/${resolution}`
    return fetch(endpoint)
      .then((res) => res.text())
      .then((text) => JSON.parse(text))
  }

  // Private methods (Requires authentication with an API key)

  async getOpenOrders(
    instrument: string,
    upto_timestamp?: number,
    page_size = 100,
  ): Promise<OpenOrder[]> {
    if (!this.credentials) {
      throw new Error("Please provide authentication credentials")
    }
    const path = "/api/v1/orders/open-orders"
    const headers = authHeaders(
      { path, method: "GET", body: "" },
      this.credentials,
    )
    if (!upto_timestamp) {
      upto_timestamp = new Date().getTime()
    }
    const endpoint =
      this.httpApi +
      `/orders/open-orders?instrument=${instrument}&upto_timestamp=${upto_timestamp}&page_size=${page_size}`
    return fetch(endpoint, { headers })
      .then((res) => res.text())
      .then((text) => JSON.parse(text))
  }

  async getPositions(): Promise<any> {
    if (!this.credentials) {
      throw new Error("Please provide authentication credentials")
    }
    const path = "/api/v1/positions"
    const headers = authHeaders(
      { path, method: "GET", body: "" },
      this.credentials,
    )
    const endpoint = this.httpApi + `/positions`
    return fetch(endpoint, { headers })
      .then((res) => res.text())
      .then((text) => JSON.parse(text))
  }

  async getAccountOverview(): Promise<any> {
    if (!this.credentials) {
      throw new Error("Please provide authentication credentials")
    }
    const path = "/api/v1/account-overview"
    const headers = authHeaders(
      { path, method: "GET", body: "" },
      this.credentials,
    )
    const endpoint = this.httpApi + `/account-overview`
    return fetch(endpoint, { headers })
      .then((res) => res.text())
      .then((text) => JSON.parse(text))
  }

  async getMyTrades(instrument: string, page?: number): Promise<any> {
    if (!this.credentials) {
      throw new Error("Please provide authentication credentials")
    }
    const path = "/api/v1/history/my-trades"
    const headers = authHeaders(
      { path, method: "GET", body: "" },
      this.credentials,
    )
    if (!page) {
      page = 0
    }
    const endpoint =
      this.httpApi + `/history/my-trades?instrument=${instrument}&page=${page}`
    return fetch(endpoint, { headers })
      .then((res) => res.text())
      .then((text) => JSON.parse(text))
  }

  placeOrder(order: Order): RequestId {
    const order_id = order.order_id || uuidv4()
    this.send({
      command: "place-order",
      ...order,
      order_id,
    })
    return order_id
  }

  cancelOrder(cancel: CancelOrder): RequestId {
    const cancel_id = cancel.cancel_id || uuidv4()
    this.send({
      command: "cancel-order",
      ...cancel,
      cancel_id,
    })
    return cancel_id
  }

  placeStopOrder(order: StopOrder): RequestId {
    const order_id = order.order_id || uuidv4()
    this.send({
      command: "place-order",
      ...order,
      order_id,
    })
    return order_id
  }

  cancelStopOrder(cancel: CancelOrder): RequestId {
    const cancel_id = cancel.cancel_id || uuidv4()
    this.send({
      command: "cancel-stop-order",
      ...cancel,
      cancel_id,
    })
    return cancel_id
  }
}
